File: samples/HelloWorld/quicklua/QRectangle.lua
--[[/*
 * (C) 2012-2013 Marmalade.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */--]]

--------------------------------------------------------------------------------
-- Rectangle
-- NOTE: This file must have no dependencies on the ones loaded after it by
-- openquick_init.lua. For example, it must have no dependencies on QDirector.lua
--------------------------------------------------------------------------------
QRectangle = {}
table.setValuesFromTable(QRectangle, QLines) -- previous class in hierarchy
QRectangle.__index = QRectangle

--------------------------------------------------------------------------------
-- Private API
--------------------------------------------------------------------------------
QRectangle.serialize = function(o)
	local obj = serializeTLMT(getmetatable(o), o)
    table.setValuesFromTable(obj, serializeTLMT(getmetatable(quick.QRectangle), o))
    table.setValuesFromTable(obj, serializeTLMT(getmetatable(quick.QLines), o))
    table.setValuesFromTable(obj, serializeTLMT(getmetatable(quick.QVector), o))
	return obj
end

--[[
/*
Initialise the peer table for the C++ class QRectangle.
This must be called immediately after the QRectangle() constructor.
*/
--]]
function QRectangle:initRectangle(n)
	local np = {}
    local ep = tolua.getpeer(n)
    table.setValuesFromTable(np, ep)
	setmetatable(np, QRectangle)
	tolua.setpeer(n, np)

    local mt = getmetatable(n) 
    mt.__serialize = QRectangle.serialize
end

--------------------------------------------------------------------------------
-- Public API
--------------------------------------------------------------------------------
--[[
/**
Create a rectangle node, specifying arbitrary input values.
@param values Lua table specifying name/value pairs.
@return The created rectangle object.
*/
--]]
function director:createRectangle(values)
end
--[[
/**
Create a rectangle node, with the specified position and radius.
@param x X coordinate of rectangle origin.
@param y Y coordinate of rectangle origin.
@param w The width of the rectangle.
@param h The height of the rectangle.
@return The created rectangle object.
*/
--]]
function director:createRectangle(x, y, w, h)
    local n = quick.QRectangle()
    -- Must call init.. on all subclasses in hierarchy order
    QNode:initNode(n)
    QVector:initVector(n)
    QLines:initLines(n)
    QRectangle:initRectangle(n)

    if (type(x) == "table") then
        table.setValuesFromTable(n, x)
    else
        dbg.assertFuncVarTypes({"number", "number", "number", "number"}, x, y, w, h)
        n.x = x
        n.y = y
        n.w = w
        n.h = h
    end

    self:addNodeToLists(n)
    return n
end

File: source/QRectangle.cpp
/*
 * (C) 2012-2013 Marmalade.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

//#include "AppDelegate.h"
#include "QRectangle.h"
#include "QLuaHelpers.h"

#include "cocos2d.h"

USING_NS_CC;

QUICK_NAMESPACE_BEGIN;

//------------------------------------------------------------------------------
// QRectangle
//------------------------------------------------------------------------------
QRectangle::QRectangle()
{
	// Cocos stuff
    // The CCSpriteLines is created by the parent constructor

    // Set up dummy points
    m_Points.push_back(QVec2::g_Zero);
    m_Points.push_back(QVec2::g_Zero);
    m_Points.push_back(QVec2::g_Zero);
    m_Points.push_back(QVec2::g_Zero);
    m_Points.push_back(QVec2::g_Zero);

    // Set up actual normals
    m_Norms.push_back(QVec2( -0.707f, -0.707f));
    m_Norms.push_back(QVec2(  0.707f, -0.707f));
    m_Norms.push_back(QVec2(  0.707f,  0.707f));
    m_Norms.push_back(QVec2( -0.707f,  0.707f));
    m_Norms.push_back(QVec2( -0.707f, -0.707f));

    // Set up actual normal lengths
    float r2 = float(sqrt(2.0f));
    m_NormLens.push_back(r2);
    m_NormLens.push_back(r2);
    m_NormLens.push_back(r2);
    m_NormLens.push_back(r2);
    m_NormLens.push_back(r2);

    // Set up GL buffers, etc.
	m_NumPoints = m_Points.size();
    _appendReallocBuffers();
}
//------------------------------------------------------------------------------
QRectangle::~QRectangle()
{
//    QTrace("QRectangle destroyed");
}
//------------------------------------------------------------------------------
void QRectangle::sync()
{
    // Only sync if sync is enabled
    if (!isSynced)
    {
        return;
    }

    QLines::sync();

    // Set points based on w, h
    m_Points[0].Set(0, 0);
    m_Points[1].Set(w, 0);
    m_Points[2].Set(w, h);
    m_Points[3].Set(0, h);
    m_Points[4].Set(0, 0);
}

QUICK_NAMESPACE_END;
